osu.Game.Rulesets.Taiko/UI/TaikoMascotTextureAnimation.cs

```diff
@@ -0,0 +1,107 @@
+﻿// Copyright (c) ppy Pty Ltd <contact@ppy.sh>. Licensed under the MIT Licence.
+// See the LICENCE file in the repository root for full licence text.
+
+using System;
+using osu.Framework.Allocation;
+using osu.Framework.Graphics;
+using osu.Framework.Graphics.Animations;
+using osu.Framework.Graphics.Textures;
+using osu.Game.Skinning;
+
+namespace osu.Game.Rulesets.Taiko.UI
+{
+    public sealed class TaikoMascotTextureAnimation : TextureAnimation
+    {
+        private const float clear_animation_speed = 1000 / 10f;
+        private static readonly int[] clear_animation_sequence = { 0, 1, 2, 3, 4, 5, 6, 5, 6, 5, 4, 3, 2, 1, 0 };
+        private int currentFrame;
+
+        public TaikoMascotAnimationState State { get; }
+
+        public TaikoMascotTextureAnimation(TaikoMascotAnimationState state)
+            : base(true)
+        {
+            State = state;
+
+            // We're animating on beat if it's not the clear animation
+            if (state == TaikoMascotAnimationState.Clear)
+                DefaultFrameLength = clear_animation_speed;
+            else
+                this.Stop();
+
+            Origin = Anchor.BottomLeft;
+            Anchor = Anchor.BottomLeft;
+        }
+
+        [BackgroundDependencyLoader]
+        private void load(ISkinSource skin)
+        {
+            if (State == TaikoMascotAnimationState.Clear)
+            {
+                foreach (var textureIndex in clear_animation_sequence)
+                {
+                    var textureName = getStateTextureName(textureIndex);
+                    Texture texture = skin.GetTexture(textureName);
+
+                    if (texture == null)
+                        break;
+
+                    AddFrame(texture);
+                }
+            }
+            else
+            {
+                for (int i = 0; true; i++)
+                {
+                    var textureName = getStateTextureName(i);
+                    Texture texture = skin.GetTexture(textureName);
+
+                    if (texture == null)
+                        break;
+
+                    AddFrame(texture);
+                }
+            }
+        }
+
+        /// <summary>
+        /// Advances the current frame by one.
+        /// </summary>
+        public void Move()
+        {
+            // Check whether there are frames before causing a crash.
+            if (FrameCount == 0)
+                return;
+
+            if (FrameCount <= currentFrame)
```

Comment:
This check reads a bit backwards - I'd expect the thing being bound-checked to be on the left side of the expression (now I have to mentally invert to make sure it's correct). You could also remove this guard entirely and incorporate modulo arithmetic when incrementing `currentFrame` at the end.

Reply:
Would this change be fine then?
```diff
- if (FrameCount <= currentFrame)
+ if (currentFrame >= FrameCount)
```

Reply:
That'd work too I suppose, yes.